File: backend/app/schemas/evaluation.py
"""应用评测请求/响应模型（08 §4.9 / 09）。"""

from __future__ import annotations

import uuid
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field


class EvalSetCreate(BaseModel):
    name: str = Field(min_length=1, max_length=100, description="名称（唯一，≤100 字）")
    description: str = Field(default="", max_length=200)


class EvalSetUpdate(BaseModel):
    name: str = Field(min_length=1, max_length=100)
    description: str = Field(default="", max_length=200)


class EvalSetOut(BaseModel):
    id: uuid.UUID
    name: str
    description: str
    source: str
    sample_count: int = 0
    created_at: datetime
    updated_at: datetime


class SampleCreate(BaseModel):
    question: str = Field(min_length=1, max_length=500)
    expected_answer: str = Field(min_length=1, max_length=2000)
    expected_chunks: list[uuid.UUID] = Field(default_factory=list)


class SampleImport(BaseModel):
    items: list[SampleCreate] = Field(min_length=1, max_length=500)


class SampleOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    eval_set_id: uuid.UUID
    question: str
    expected_answer: str
    expected_chunks: list
    source: str
    created_at: datetime


class EvalTaskCreate(BaseModel):
    eval_set_id: uuid.UUID
    model_profile_id: uuid.UUID | None = None
    kb_ids: list[uuid.UUID] = Field(min_length=1, max_length=20, description="评测检索知识库")


class EvalTaskOut(BaseModel):
    id: uuid.UUID
    eval_set_id: uuid.UUID
    eval_set_name: str = ""
    model_profile_id: uuid.UUID | None = None
    model_name: str = ""
    kb_ids: list
    status: str
    progress: int
    total: int
    score_avg: Decimal | None = None
    metrics: dict | None = None
    error_message: str | None = None
    created_at: datetime
    updated_at: datetime


class EvalResultOut(BaseModel):
    id: uuid.UUID
    sample_id: uuid.UUID
    question: str
    expected_answer: str
    answer: str
    citations: list
    scores: dict
    passed: bool


class EvalReportOut(BaseModel):
    task: EvalTaskOut
    score_avg: Decimal | None = None
    pass_rate: float = 0.0
    total: int = 0
    passed_count: int = 0
    metrics: dict | None = None
    results: list[EvalResultOut] = []


class PassUpdate(BaseModel):
    passed: bool


class CandidateOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    question: str
    expected_answer: str
    source: str
    source_id: str | None = None
    message_id: uuid.UUID | None = None
    status: str
    created_at: datetime


class CandidateConfirm(BaseModel):
    eval_set_id: uuid.UUID

File: frontend/src/api/evaluation.ts
/** 应用评测接口（B4.5）。 */

import { request } from './client';

export interface EvalSet {
  id: string;
  name: string;
  description: string;
  source: string;
  sample_count: number;
  created_at: string;
  updated_at: string;
}

export interface EvalSample {
  id: string;
  eval_set_id: string;
  question: string;
  expected_answer: string;
  expected_chunks: string[];
  source: string;
  created_at: string;
}

export interface EvalTask {
  id: string;
  eval_set_id: string;
  eval_set_name: string;
  model_profile_id: string | null;
  model_name: string;
  kb_ids: string[];
  status: 'pending' | 'running' | 'completed' | 'failed';
  progress: number;
  total: number;
  score_avg: number | null;
  metrics: { accuracy?: number; pass_rate?: number; passed_count?: number } | null;
  error_message: string | null;
  created_at: string;
  updated_at: string;
}

export interface EvalResult {
  id: string;
  sample_id: string;
  question: string;
  expected_answer: string;
  answer: string;
  citations: Array<Record<string, unknown>>;
  scores: { accuracy: number | null; relevancy: number | null; semantic: number | null };
  passed: boolean;
}

export interface EvalReport {
  task: EvalTask;
  score_avg: number | null;
  pass_rate: number;
  total: number;
  passed_count: number;
  metrics: { accuracy?: number; pass_rate?: number; passed_count?: number } | null;
  results: EvalResult[];
}

export interface EvalCandidate {
  id: string;
  question: string;
  expected_answer: string;
  source: string;
  source_id: string | null;
  message_id: string | null;
  status: string;
  created_at: string;
}

export interface PageResult<T> {
  items: T[];
  total: number;
  page: number;
  page_size: number;
}

// ---------- 评测集 ----------

export function listEvalSets(): Promise<EvalSet[]> {
  return request({ url: '/evaluations/sets', method: 'GET' });
}

export function createEvalSet(data: { name: string; description: string }): Promise<EvalSet> {
  return request({ url: '/evaluations/sets', method: 'POST', data });
}

export function updateEvalSet(
  id: string,
  data: { name: string; description: string },
): Promise<EvalSet> {
  return request({ url: `/evaluations/sets/${id}`, method: 'PUT', data });
}

export function deleteEvalSet(id: string): Promise<null> {
  return request({ url: `/evaluations/sets/${id}`, method: 'DELETE' });
}

// ---------- 样本 ----------

export function listSamples(
  setId: string,
  params: { page: number; page_size: number },
): Promise<PageResult<EvalSample>> {
  return request({ url: `/evaluations/sets/${setId}/samples`, method: 'GET', params });
}

export function addSample(
  setId: string,
  data: { question: string; expected_answer: string; expected_chunks?: string[] },
): Promise<EvalSample> {
  return request({ url: `/evaluations/sets/${setId}/samples`, method: 'POST', data });
}

export function importSamples(
  setId: string,
  items: { question: string; expected_answer: string; expected_chunks?: string[] }[],
): Promise<null> {
  return request({ url: `/evaluations/sets/${setId}/samples/import`, method: 'POST', data: { items } });
}

export function importPublicSamples(setId: string): Promise<null> {
  return request({ url: `/evaluations/sets/${setId}/samples/import-public`, method: 'POST' });
}

// ---------- 任务 ----------

export function listTasks(params: { page: number; page_size: number }): Promise<PageResult<EvalTask>> {
  return request({ url: '/evaluations/tasks', method: 'GET', params });
}

export function createTask(data: {
  eval_set_id: string;
  model_profile_id?: string | null;
  kb_ids: string[];
}): Promise<EvalTask> {
  return request({ url: '/evaluations/tasks', method: 'POST', data });
}

export function getTask(id: string): Promise<EvalTask> {
  return request({ url: `/evaluations/tasks/${id}`, method: 'GET' });
}

export function rerunTask(id: string): Promise<EvalTask> {
  return request({ url: `/evaluations/tasks/${id}/rerun`, method: 'POST' });
}

export function deleteTask(id: string): Promise<null> {
  return request({ url: `/evaluations/tasks/${id}`, method: 'DELETE' });
}

export function getReport(id: string): Promise<EvalReport> {
  return request({ url: `/evaluations/tasks/${id}/report`, method: 'GET' });
}

export function updateResultPassed(id: string, passed: boolean): Promise<null> {
  return request({ url: `/evaluations/results/${id}/passed`, method: 'PUT', data: { passed } });
}

// ---------- 回流候选 ----------

export function listCandidates(): Promise<EvalCandidate[]> {
  return request({ url: '/evaluations/candidates', method: 'GET' });
}

export function confirmCandidate(id: string, evalSetId: string): Promise<null> {
  return request({ url: `/evaluations/candidates/${id}/confirm`, method: 'POST', data: { eval_set_id: evalSetId } });
}

export function rejectCandidate(id: string): Promise<null> {
  return request({ url: `/evaluations/candidates/${id}/reject`, method: 'POST' });
}

